_posts/2016/2016-12-20-cqrs-and-elasticsearch.md

```diff
@@ -0,0 +1,228 @@
+
+
+## Partial Updates of Documents
+
+[Updating][es-update] an already indexed complete document in Elasticsearch is pretty straight-forward:
+
+{% highlight javascript %}
+PUT /products_12345/product/456
+{
+    "sku" : "334163-2",
+    "name" : "Sneakers",
+    "description" : "best running experience",
+    "price" : 49.99,
+    "availability" : "LOW_STOCK",
+    "images" : [{
+        "name" : "front",
+        "uri" : "https://myshop.com/images/334163-2/front.jpg"
+    }, {
+        "name" : "back",
+        "uri" : "https://myshop.com/images/334163-2/back.jpg"
+    }]
+}
+{% endhighlight %}
+
+But when the update is triggered by one of the aforementioned events, we need to be a bit more careful:
+When processing an *availability-updated-event*, just computing the new *availability* state and including it as the only attribute in the update would wipe out all the other attributes of this product!
+
+{% highlight javascript %}
+PUT /products_12345/product/456
+{
+    "availability" : "LOW_STOCK"
+}
+
+GET /products_12345/product/456
+{
+    "_index" : "products_12345",
+    "_type" : "product",
+    "_id" : "456",
+    "_version" : 2,
+    "found" : true,
+    "_source" : {
+        "availability" : "LOW_STOCK",
+    }
+}
+{% endhighlight %}
+
+Luckily Elasticsearch offers a feature called [partial updates][es-partial-update] to mitigate exactly this problem, leaving all remaining attributes untouched:
+
+{% highlight javascript %}
+POST /products_12345/product/456/_update
+{
+    "doc" :  {
+        "availability" : "LOW_STOCK"
+    }
+}
+
+GET /products_12345/product/456
+{
+    "_index" : "products_12345",
+    "_type" :  "product",
+    "_id" :    "456",
+    "_version" : 2,
+    "found" : true,
+    "_source" : {
+        "sku" : "334163-2",
+        "name" : "Sneakers",
+        "description" : "best running experience",
+        "price" : 49.99,
+        "availability" : "LOW_STOCK",
+        "images" : [{
+            "name" : "front",
+            "uri" : "https://myshop.com/images/334163-2/front.jpg"
+        }, {
+            "name" : "back",
+            "uri" : "https://myshop.com/images/334163-2/back.jpg"
+        }]
+    }
+}
+{% endhighlight %}
+
+## Conclusion
+
+Keeping decentrally managed data in sync is not trivial.
+Even when accepting [Eventual Constistency][fowler-eventual-consistency], you need to implement data synchronization processes and take care of storage-specific peculiarities.
+
+Asynchronous processing of events imposes another complexity by having no guarantees on the order of received messages.
+Introducing Elasticsearch [optimistic concurrency control][es-concurrency] and retrying out-of-order events could minimize this, at the cost of yet another complex implementation detail.
+
+The advantages of solving all these challenges must justify the effort involved.
+In our microservices architecture we already rely on messaging for data synchronization, so there is no additional complexity introduced.
```

Comment:
That's not the idea. Data Sync is complex by nature, but that is complexity we already have. Separating product-mgmt from product-view does not add to the already incurred complexity.